[MASTG-TOOL-0015] Update drozer content and remove outdated references (#2614)
OWASP/owasp-mastg: tools/android/MASTG-TOOL-0015.md

---
title: drozer
platform: android
source: https://github.com/WithSecureLabs/drozer
---

> [drozer](https://github.com/WithSecureLabs/drozer "drozer on GitHub") a security testing framework for Android. that allows you to search for security vulnerabilities in apps and devices by assuming the role of an app interacting with Android Runtime, other apps' IPC endpoints, and the underlying OS.

Reviewer: ```suggestion
[drozer](https://github.com/WithSecureLabs/drozer "drozer on GitHub") is a security testing framework for Android that allows you to search for security vulnerabilities in apps and devices by assuming the role of an app interacting with the Android Runtime, other apps' IPC endpoints, and the underlying OS.
```

drozer helps to reduce the time taken for Android security assessments by automating tedious and time-consuming tasks. It allows testers and reverse-engineers to:

- Discover and interact with the attack surface exposed by Android apps.
- Execute dynamic Java-code on a device, to avoid the need to compile and install small test scripts.

> drozer runs both in Android emulators and on real devices. It does not require USB debugging or other development features to be enabled; so you can perform assessments on devices in their production state to get better results.

Reviewer: Agree with this change. You also don't need root for ADB and you don't even need ADB to install additional APKs.

First, you need to set up a suitable port forward so that your PC can connect to a TCP socket opened by the Agent inside the emulator, or on the device. By default, drozer uses port 31415

```
adb forward tcp:31415 tcp:31415
```

You can now connect to the drozer Agent by connecting to `localhost` (or simply not specifying the target IP)

```
drozer console connect
```

### Confirming a successful connection

You should be presented with a drozer command prompt:

```
Selecting ebe9fcc0c47b28da (Google sdk_gphone64_x86_64 12)

            ..                    ..:.
           ..o..                  .r..
            ..a..  . ....... .  ..nd
              ro..idsnemesisand..pr
              .otectorandroidsneme.
           .,sisandprotectorandroids+.
         ..nemesisandprotectorandroidsn:.
        .emesisandprotectorandroidsnemes..
      ..isandp,..,rotecyayandro,..,idsnem.
      .isisandp..rotectorandroid..snemisis.
      ,andprotectorandroidsnemisisandprotec.
     .torandroidsnemesisandprotectorandroid.
     .snemisisandprotectorandroidsnemesisan:
     .dprotectorandroidsnemesisandprotector.

drozer Console (v3.0.0)
dz>
```

> The prompt confirms the Android ID of the device you have connected to, along with the manufacturer, model and Android software version.

Reviewer: ```suggestion
The prompt confirms the Android ID of the device you have connected to, along with the manufacturer, model and Android OS version.
```

Some common drozer commands include:

#### Searching for applications on the device:
```
run app.package.list -f <keyword>
```
This lists basic informations about any packages containing the word "<keyword>" in their bundle identifier. This includes package names, key directories used by the application, and any permissions used or defined by the application.

#### Enumerating the attack surface of an app:
```
run app.package.attacksurface <package>
```
This command inspects the target app's manifest and provides a report on any exported components of the application, and verifies whether the application is debuggable.

Once the attack surface has been identified, you can obtain more specific information about each component class. For example, to list  Activities, you can use the following command:
```
run app.activity.info -a <package>
```
This lists the names of all exported Activities, together with the permissions required to interact with them.

#### Starting an Activity
In order to launch an exported activity, use the following command:
```
run app.activity.start --component <package> <component name>
```

> When calling `app.activity.start`, you can build a much more complex intent. As with all drozer modules, you can request more usage information:

Reviewer 2: ```suggestion
When calling `app.activity.start`, you can build a much more complex intent. As with all drozer modules, you can request more usage information by using the command `help`:
```

```
dz> help app.activity.start
Attempting to run shell module
usage: run app.activity.start [-h] [--action ACTION] [--category CATEGORY [CATEGORY ...]] [--component PACKAGE COMPONENT] [--data-uri DATA_URI] [--extra TYPE KEY VALUE] [--flags FLAGS [FLAGS ...]] [--mimetype MIMETYPE]
```

> You can learn more about how intents are created by running `help intents`

Reviewer 2: ```suggestion
You can learn more about how intents are created by running `help intents`.
```

> #### Further information

Reviewer 2: This link is already used in "### Example usage". Can you remove this short section and merge it with "### Example usage"?

Refer to the [Official drozer User Manual](https://labs.withsecure.com/tools/drozer "drozer User Manual") for a more comprehensive list of guided examples.

## Other drozer resources

Other resources where you might find useful information are:

- [Official drozer User Manual](https://labs.withsecure.com/tools/drozer "drozer User Manual")
- [drozer GitHub page](https://github.com/WithSecureLabs/drozer "GitHub repo - drozer")

> - [drozer Agent GitHub page](https://github.com/WithSecureLabs/drozer "GitHub repo - drozer-agent")

Reviewer 2: This is the same URL as the link before. I would suggest to remove it.

Author: Whoops, that second link is supposed to be https://github.com/WithSecureLabs/drozer-agent - will correct soon!

> - [drozer Agent GitHub page](https://github.com/WithSecureLabs/drozer "GitHub repo - drozer-agent")

Reviewer 3: ```suggestion
- [drozer Agent GitHub page](https://github.com/WithSecureLabs/drozer-agent "GitHub repo - drozer-agent")
```

---
title: drozer
platform: android
source: https://github.com/WithSecureLabs/drozer
---

[drozer](https://github.com/WithSecureLabs/drozer "drozer on GitHub") is a security testing framework for Android that allows you to search for security vulnerabilities in apps and devices by assuming the role of an app interacting with the Android runtime, other apps' IPC endpoints, and the underlying OS.

drozer can be used during Android security assessments to automate tasks. It allows testers and reverse engineers to:

- Discover and interact with the attack surface exposed by Android apps.
- Execute dynamic Java-code on a device, to avoid the need to compile and install small test scripts.

drozer runs both in Android emulators and on real devices. It does not require USB debugging or other development features to be enabled; so you can perform assessments on devices in their production state to simulate attacks.

You can extend drozer with additional modules to find, test and exploit other weaknesses; this, combined with scripting possibilities, helps to automate regression testing for security issues.

## Installing drozer and Setup

> Detailed instructions on how to install and setup the drozer console on your machine and the drozer agent on the Android phone can be found in the [drozer Github repo](https://github.com/WithSecureLabs/drozer "Installation instructions of drozer").

Reviewer: ```suggestion
Detailed instructions on how to install and set up the drozer console on your machine and the drozer agent on the Android phone can be found in the [drozer Github repo](https://github.com/WithSecureLabs/drozer "Installation instructions of drozer").
```

### Example usage

Once drozer is set up, you can use drozer to perform reconnaissance and exploitation of Android applications from the perspective of a malicious app on the device. [The drozer User Manual](https://labs.withsecure.com/tools/drozer#3 "drozer User Manual") introduces an intentionally vulnerable application - [sieve](https://github.com/WithSecureLabs/sieve "GitHub repo - sieve") - together with step-by-step exploitation instructions.

Some common drozer commands include:

#### Searching for applications on the device

```
run app.package.list -f <keyword>
```

> This lists basic informations about any packages containing the word "<keyword>" in their bundle identifier. This includes package names, key directories used by the application, and any permissions used or defined by the application.

Reviewer: ```suggestion
This lists basic information about any packages containing the word "<keyword>" in their package name. This includes the app name, key directories used by the application, and any permissions used or defined by the application.
```
Bundle name is only used on iOS IMO.

> This lists basic informations about any packages containing the word "<keyword>" in their bundle identifier. This includes package names, key directories used by the application, and any permissions used or defined by the application.

Reviewer 3: ```suggestion
This lists basic information about any packages containing the word "<keyword>" in their bundle identifier. This includes package names, key directories used by the application, and any permissions used or defined by the application.
```

#### Enumerating the attack surface of an app

> ```

Reviewer 3: ```suggestion
```sh
```

run app.package.attacksurface <package>
```

This command inspects the target app's manifest and provides a report on any exported components of the application, and verifies whether the application is debuggable.

Once the attack surface has been identified, you can obtain more specific information about each component class. For example, to list  Activities, you can use the following command:

> ```

Reviewer 3: ```suggestion
```sh
```

run app.activity.info -a <package>
```

This lists the names of all exported Activities, together with the permissions required to interact with them.

#### Starting an Activity

In order to launch an exported activity, use the following command:

> ```

Reviewer 3: ```suggestion
```sh
```

run app.activity.start --component <package> <component name>
```

When calling `app.activity.start`, you can build a much more complex intent. As with all drozer modules, you can request more usage information by using the `help` command:

> ```

Reviewer 3: ```suggestion
```sh
```
